src/docfx/restore/Restore.cs

```diff
@@ -41,181 +37,41 @@ internal static class Restore
-            var workTreeHeads = new ConcurrentBag<(string href, string head)>();
-
-            await ProcessUtility.CreateFileMutex(
-                PathUtility.NormalizeFile(Path.GetRelativePath(AppData.RestoreDir, restorePath)),
-                async () =>
-                {
-                    await FetchOrCloneRepo();
-
-                    var (existingWorkTrees, newWorkTrees) = await AddWorkTrees();
-
-                    if (NeedCleanupWorkTrees(existingWorkTrees.Count))
-                    {
-                        using (Progress.Start($"Cleanup `{restoreDir}` work trees"))
-                        {
-                            await CleanupWorkTrees(existingWorkTrees, newWorkTrees);
-                        }
-                    }
-                });
 
-            async Task FetchOrCloneRepo()
+            // restore git dependnecy repositories
+            var workTreeHeadMappings = await RestoreGit.Restore(docsetPath, config, restoreChild);
+            foreach (var (href, workTreeHead) in workTreeHeadMappings)
             {
-                if (GitUtility.IsRepo(restoreDir))
-                {
-                    // already exists, just pull the new updates from remote
-                    await GitUtility.Fetch(restorePath);
-                }
-                else
-                {
-                    // doesn't exist yet, clone this repo to a specified branch
-                    await GitUtility.Clone(restoreDir, url, restorePath, null, true);
-                }
+                result.Git[href] = workTreeHead;
             }
 
-            async Task<(List<string> existingWorkTrees, List<string> newWorkTrees)> AddWorkTrees()
+            // restore github user
+            if (!string.IsNullOrEmpty(config.Contribution.UserProfileCacheUrl))
             {
-                var existingWorkTrees = new ConcurrentDictionary<string, int>((await GitUtility.ListWorkTrees(restorePath, false)).ToDictionary(k => k, v => 0));
-                var newWorkTrees = new ConcurrentBag<string>();
-                await ParallelUtility.ForEach(hrefs, async href =>
-                {
-                    var (_, rev) = GetGitRemoteInfo(href);
-                    var workTreeHead = await GitUtility.Revision(restorePath, rev);
-                    var workTreePath = GetRestoreWorkTreeDir(restoreDir, workTreeHead);
-                    if (existingWorkTrees.TryAdd(workTreePath, 0))
-                    {
-                        await GitUtility.AddWorkTree(restorePath, workTreeHead, workTreePath);
-                    }
-
-                    newWorkTrees.Add(workTreePath);
-                    workTreeHeads.Add((href, workTreeHead));
-                });
-
-                return (existingWorkTrees.Keys.ToList(), newWorkTrees.ToList());
+                result.Url[config.Contribution.UserProfileCacheUrl] = await RestoreUrl.Restore(docsetPath, config.Contribution.UserProfileCacheUrl);
```

Comment:
>await [](start = 70, length = 5)

these restore urls should be restored in parallel. we may have a lot of urls in the future

Reply:
Sure

---
In reply to: [202252529](https://github.com/dotnet/docfx/pull/3065#discussion_r202252529) [](ancestors = 202252529)